[DOCS] Adds description of analysis_stats object and its properties to GET DFA stats API docs (#53881)
elastic/elasticsearch: docs/reference/ml/ml-shared.asciidoc

hyperparameter optimization steps in the Bayesian optimization procedure. 
The maximum number of steps is determined based on the number of undefined hyperparameters 
times the maximum optimization rounds per hyperparameter.
end::dfas-max-optimization-rounds[]

tag::dfas-max-trees[]
`max_trees`::::
(integer)
The maximum number of trees in the forest.
end::dfas-max-trees[]

tag::dfas-num-folds[]
`num_folds`::::
(integer)
The maximum number of folds for the cross-validation procedure.
end::dfas-num-folds[]

tag::dfas-num-splits[]
`num_splits_per_feature`::::
(integer)
Determines the maximum number of splits for every feature that can occur in a 
decision tree when the tree is trained.
end::dfas-num-splits[]

tag::dfas-soft-limit[]
`soft_tree_depth_limit`::::
(double)
Tree depth limit is used for calculating the tree depth penalty. This is a soft 
limit, it can be exceeded.
end::dfas-soft-limit[]

tag::dfas-soft-tolerance[]
`soft_tree_depth_tolerance`::::
(double)
Tree depth tolerance is used for calculating the tree depth penalty. This is a 
soft limit, it can be exceeded.
end::dfas-soft-tolerance[]
======
//End class_hyperparameters

tag::dfas-iteration[]
`iteration`::::
(integer)

> The number of iterations on the {dfanalytics-job}.

Reviewer: I would rather say here "number of iterations in the classification analysis" instead of the DFA job, as iterations is not a concept that is relevant for all analyses.

end::dfas-iteration[]

tag::dfas-timestamp[]
`timestamp`::::
(date)

> The timestamp when the statistics returned in milliseconds since the epoch.

Reviewer: I wonder if we could substitue `returned` with a more suitable verb here. Perhaps `... when the statistics were reported`?

end::dfas-timestamp[]

//Begin class_timing_stats
tag::dfas-timing-stats[]
`timing_stats`::::
(object)
An object containing time statistics about the {dfanalytics-job}.
end::dfas-timing-stats[]
+
.Properties of `timing_stats`
[%collapsible%open]
======
tag::dfas-timing-stats-elapsed[]
`elapsed_time`::::
(integer)

> Runtime of the {dfanalytics-job} in milliseconds.

Reviewer: This should explain this is the runtime of the analysis itself, not the whole job's. Note the job involves other steps that take time (e.g. reindexing, etc.) that are not accounted for by this measurement.

end::dfas-timing-stats-elapsed[]

tag::dfas-timing-stats-iteration[]
`iteration_time`::::
(integer)

> Runtime of the {dfanalytics-job} since the last iteration in milliseconds.

Reviewer: It might be simpler to explain this as "Runtime of the latest iteration of the classification analysis in milliseconds"

end::dfas-timing-stats-iteration[]
======
//End class_timing_stats

//Begin class_validation_loss
tag::dfas-validation-loss[]
`validation_loss`::::
(object)
An object containing information about validation loss.
end::dfas-validation-loss[]
+
.Properties of `validation_loss`
[%collapsible%open]
======
tag::dfas-validation-loss-type[]
`loss_type`::::
(string)

> The name of the loss metric. For example, `binomial_logistic`.

Reviewer: `name` -> `type` ?

An array of statistics objects for {dfanalytics-jobs}, which are
sorted by the `id` value in ascending order.

//Begin analysis_stats
`analysis_stats`::
(object)
An object containing statistical data about the analysis.
+
.Properties of `analysis_stats`
[%collapsible%open]
====
//Begin classification_stats
`classification_stats`:::
(object)
An object containing statistical data about the {classanalysis}.
+
.Properties of `classification_stats`
[%collapsible%open]
=====
//Begin class_hyperparameters
`hyperparameters`::::
(object)
An object containing the parameters of the {classanalysis}.
+
.Properties of `hyperparameters`
[%collapsible%open]
======
`alpha`::::
(double)
Regularization factor to penalize deeper trees when training decision trees.

`class_assignment_objective`::::
(string)
Defines whether class assignment maximizes the accuracy or the minimum recall 
metric. Possible values are `maximize_accuracy` and `maximize_minimum_recall`.

`downsample_factor`::::
(double)
The value of the downsample factor.

`eta`::::
(double)
The value of the eta hyperparameter.

`eta_growth_rate_per_tree`::::
(double)

> Specifies with what rate to increase `eta` for every new tree added to the 

Reviewer 2: Minor suggestion for clarity:
```suggestion
Specifies the rate at which the `eta` increases for each new tree that is added to the 
```

> forest. The rate of 1.0 does not change `eta`, the rate of 1.05 increases `eta` 

Reviewer 2: I think you can likely simplify this as follows:
```suggestion
forest. For example, a rate of `1.05` increases `eta` 
```

by 5%.

`feature_bag_fraction`::::
(double)
The fraction of features that is used when selecting a random bag for each 
candidate split.

`gamma`::::
(double)
Regularization factor to penalize trees with large numbers of nodes.

`lambda`::::
(double)
Regularization factor to penalize large leaf weights.

`max_attempts_to_add_tree`::::
(integer)
If the algorithm fails to determine a non-trivial tree (more than a single 
leaf), this parameter determines how many of such consecutive failures are 
tolerated. Once the number of attempts exceeds the threshold, the forest 
training stops.

`max_optimization_rounds_per_hyperparameter`::::
(integer)
A multiplier responsible for determining the maximum number of 

> hyperparameter optimization steps in the bayesian optimization procedure. 

Reviewer 2: I think this should be capitalized:
```suggestion
hyperparameter optimization steps in the Bayesian optimization procedure. 
```

> Maximum number of steps is determined as the number of undefined hyperparameters 

Reviewer 2: ```suggestion
The maximum number of steps is determined based on the number of undefined hyperparameters 
```

times the maximum optimization rounds per hyperparameter.

`max_trees`::::
(integer)
The maximum number of trees in the forest.

`num_folds`::::
(integer)

> Maximum number of folds for the cross-validation procedure.

Reviewer 2: Synching with definition above this one:
```suggestion
The maximum number of folds for the cross-validation procedure.
```

`distance_kth_nn`, `distance_knn`, and `ensemble`.

`n_neighbors`::::
(integer)
The value for how many nearest neighbors each method of {oldetection} uses to 
calculate its outlier score.

`outlier_fraction`::::
(double)
The proportion of the data set that is assumed to be outlying prior to 
{oldetection}.

`standardization_enabled`::::
(boolean)
If true, then the following operation is performed on the columns before 
computing {olscores}: (x_i - mean(x_i)) / sd(x_i).
======
//End parameters

`timestamp`::::
(date)
The timestamp when the statistics returned in milliseconds since the epoch.

//Begin od_timing_stats
`timing_stats`::::
(object)
An object containing time statistics about the {dfanalytics-job}.
+
.Property of `timing_stats`
[%collapsible%open]
======
`elapsed_time`::::
(integer)
Runtime of the {dfanalytics-job} in milliseconds.
======
//End od_timing_stats
=====
//End outlier_detection_stats

//Begin regression_stats
`regression_stats`:::
(object)
An object containing statistical data about the {reganalysis}.
+
.Properties of `regression_stats`
[%collapsible%open]
=====
//Begin reg_hyperparameters
`hyperparameters`::::
(object)

> An object containing the paramters of the {reganalysis}.

Reviewer 2: ```suggestion
An object containing the parameters of the {reganalysis}.
```

+
.Properties of `hyperparameters`
[%collapsible%open]
======
`alpha`::::
(double)
Regularization factor to penalize deeper trees when training decision trees.

`downsample_factor`::::
(double)
The value of the downsample factor.

`eta`::::
(double)
The value of the eta hyperparameter.

`eta_growth_rate_per_tree`::::
(double)

> Specifies with what rate to increase `eta` for every new tree added to the 

Reviewer 2: Clarifying to match earlier similar definition:
```suggestion
Specifies the rate at which the `eta` increases for every new tree that is added to the 
```

forest. The rate of 1.0 does not change `eta`, the rate of 1.05 increases `eta` 
by 5%.

`feature_bag_fraction`::::
(double)
The fraction of features that is used when selecting a random bag for each 
candidate split.

`gamma`::::
(double)
Regularization factor to penalize trees with large numbers of nodes.

`lambda`::::
(double)
Regularization factor to penalize large leaf weights.

`max_attempts_to_add_tree`::::
(integer)
If the algorithm fails to determine a non-trivial tree (more than a single 
leaf), this parameter determines how many of such consecutive failures are 
tolerated. Once the number of attempts exceeds the threshold, the forest 
training stops.

`max_optimization_rounds_per_hyperparameter`::::
(integer)

> A multiplier responsible for determining the maximum number of 

Reviewer 2: > A multiplier responsible ... 

This content seems to duplicate an earlier definition.  Can the content be re-used?

Author: @lcawl Yes, it can be. I single-sourced all the possible shared properties in a7bec16.

An array of statistics objects for {dfanalytics-jobs}, which are
sorted by the `id` value in ascending order.

//Begin analysis_stats
`analysis_stats`::
(object)
An object containing statistical data about the analysis.
+
.Properties of `analysis_stats`
[%collapsible%open]
====
//Begin classification_stats
`classification_stats`:::
(object)
An object containing statistical data about the {classanalysis}.
+
.Properties of `classification_stats`
[%collapsible%open]
=====
//Begin class_hyperparameters
`hyperparameters`::::
(object)

> An object containing the hyperparameters of the {classanalysis}.

Reviewer 3: ```suggestion
An object containing the parameters of the {classanalysis}.
```
Also the section is called "hyperparameters" not all of these parameters are estimated in a hyperparameter  optimization procedure. Hence, I would use the word "parameters" here.

+
.Properties of `hyperparameters`
[%collapsible%open]
======
`class_assignment_objective`::::
(string)
Defines whether class assignment maximizes the accuracy or the minimum recall 

> metric. Possible values are `accuracy` and `minimum_recall`.

Reviewer 3: ```suggestion
metric. Possible values are `maximize_accuracy` and `maximize_minimum_recall`.
```
Schema will be updated.

`downsample_factor`::::
(double)
The value of the downsample factor.

`eta`::::
(double)

> The value of the eta hyperparameter.

Reviewer 3: Since `eta` can be specified explicitly in the request body, maybe you can cross-reference it here?

`eta_growth_rate_per_tree`::::
(double)
Specifies with what rate to increase `eta` for every new tree added to the 
forest. The rate of 1.0 does not change `eta`, the rate of 1.05 increases `eta` 
by 5%.

`feature_bag_fraction`::::
(double)
The fraction of features that is used when selecting a random bag for each 
candidate split.

`max_attempts_to_add_tree`::::
(integer)
If the algorithm fails to determine a non-trivial tree (more than a single 
leaf), this parameter determines how many of such consecutive failures are 
tolerated. Once the number of attempts exceeds the threshold, the forest 
training stops.

`max_optimization_rounds_per_hyperparameter`::::
(integer)
A multiplier responsible for determining the maximum number of 
hyperparameter optimization steps in the bayesian optimization procedure. 
Maximum number of steps is determined as the number of undefined hyperparameters 
times the maximum optimization rounds per hyperparameter.

`max_trees`::::
(integer)

> The maximum number of trees.

Reviewer 3: ```suggestion
The maximum number of trees in the forest.
```

`num_folds`::::
(integer)
Maximum number of folds for the cross-validation procedure.

`num_splits_per_feature`::::
(integer)
Determines the maximum number of splits for every feature that can occur in a 
decision tree when the tree is trained.

> `regularization_depth_penalty_multiplier`::::

Reviewer 3: After https://github.com/elastic/ml-cpp/pull/1096 is through, it will be called `alpha`, it can be cross-referenced with analysis config.

(double)
Regularization factor to penalize deeper trees when training decision trees.

> `regularization_leaf_weight_penalty_multiplier`::::

Reviewer 3: This guy will be simply called `lambda`

(double)
Regularization factor to penalize large leaf weights. This parameter is also 
referred to as `lambda`.

`regularization_soft_tree_depth_limit`::::
(double)
Tree depth limit is used for calculating the tree depth penalty. This is a soft 
limit, it can be exceeded.

`regularization_soft_tree_depth_tolerance`::::
(double)
Tree depth tolerance is used for calculating the tree depth penalty. This is a 
soft limit, it can be exceeded.

> `regularization_tree_size_penalty_multiplier`::::

Reviewer 3: This parameter will be called `gamma`, the same as in the analysis config.
